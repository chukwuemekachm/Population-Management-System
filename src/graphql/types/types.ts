import { Request } from 'express-serve-static-core';
import { GraphQLResolveInfo } from 'graphql';

import { Prisma } from '../prisma/generated/prisma-client';

// Data Types
export type DateTime = string;
export type ID = string;
export declare type ResolverFn = <T>(
  parent?: T,
  args?: T,
  context?: T,
  info?: GraphQLResolveInfo,
) => AsyncIterator<T>;

// Resolver Types
export type AuthResponseResolver = (parent: AuthResponse) => AuthResponse;
export type UserResolver = (parent: User) => User;
export type LocationResolver = (parent: Location) => Location;

// Common Types
export interface Context {
  request: Request | { get: () => string };
  prisma: Prisma;
  user?: Partial<User>;
}

export interface AuthResponse {
  token: string;
  user: Partial<User>;
}

// Model Types
export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  locationName: string;
  malePopulation: number;
  femalePopulation: number;
  totalPopulation: number;
  parentLocation: Location | null;
  subLocations: Location[];
  creator: User | null;
  createdAt: DateTime;
  updatedAt: DateTime;
}
