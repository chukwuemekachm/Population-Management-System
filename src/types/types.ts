import { Request } from 'express-serve-static-core';

import { Prisma } from '../prisma/generated/prisma-client';

export type DateTime = string;
export type ID = string;

export interface Context {
  request: Request;
  prisma: Prisma;
}

export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  name: string;
  maleResidents: number;
  femaleResidents: number;
  totalResidents: number;
  parentLocationId: ID;
  subLocations: Location[];
  creator: User;
  createdAt: DateTime;
  updatedAt: DateTime;
}
