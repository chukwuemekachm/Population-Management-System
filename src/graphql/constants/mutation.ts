import { gql } from 'apollo-server-express';

export const SIGNUP_MUTATION = gql`
  mutation signup($user: SignupInput) {
    signup(user: $user) {
      user {
        email
        firstName
        lastName
      }
    }
  }
`;

export const LOGIN_MUTATION = gql`
  mutation login($user: LoginInput) {
    login(user: $user) {
      user {
        firstName
        lastName
      }
    }
  }
`;

export const CREATE_LOCATION_MUTATION = gql`
  mutation login($location: LocationInput) {
    createLocation(location: $location) {
      locationName
      malePopulation
      femalePopulation
    }
  }
`;
