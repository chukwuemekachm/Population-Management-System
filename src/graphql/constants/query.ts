import { gql } from 'apollo-server-express';

export const GET_LOCATIONS_QUERY = gql`
  query getLocations($filter: LocationFilterInput) {
    getLocations(filter: $filter) {
      locationName
    }
  }
`;
