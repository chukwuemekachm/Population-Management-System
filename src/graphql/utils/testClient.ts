import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';

import typeDefs from '../schema';
import resolvers from '../resolvers';
import prismaTestMockClient from '../__mocks__/prisma';
import { User } from '../types/types';

export const constructGraphqlTestClient = (user?: Partial<User>) => {
  const testServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: ({ req }) => ({
      request: req,
      prisma: prismaTestMockClient,
      user,
    }),
  });

  return createTestClient(testServer);
};
