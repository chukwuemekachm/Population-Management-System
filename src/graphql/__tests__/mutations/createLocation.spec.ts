import prismaTestMockClient from '../../__mocks__/prisma';
import { CREATE_LOCATION_MUTATION } from '../../constants';
import { User } from '../../types/types';
import { locationInputs } from '../../fixtures/locationFixtures';
import { userInputs } from '../../fixtures/userFixtures';
import { constructGraphqlTestClient } from '../../utils/testClient';

let user: Partial<User>;

describe('create location mutation', () => {
  beforeAll(async () => {
    await prismaTestMockClient.deleteManyUsers();
    user = await prismaTestMockClient.createUser(userInputs[1]);
  });
  afterAll(async () => {
    await prismaTestMockClient.deleteManyUsers();
  });

  test('should return newly created location', async () => {
    const { mutate } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        location: locationInputs[1],
      },
    };
    const { errors, data } = await mutate({
      mutation: CREATE_LOCATION_MUTATION,
      ...payload,
    });
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  test('should return validation errors when inputs are not provided', async () => {
    const { mutate } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        location: locationInputs[0],
      },
    };
    const { errors, data } = await mutate({
      mutation: CREATE_LOCATION_MUTATION,
      ...payload,
    });
    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors).toMatchSnapshot();
  });

  test('should return a location exists error', async () => {
    const { mutate } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        location: locationInputs[1],
      },
    };
    const { errors, data } = await mutate({
      mutation: CREATE_LOCATION_MUTATION,
      ...payload,
    });
    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors).toMatchSnapshot();
  });

  test('should return an invalid parentId error', async () => {
    const { mutate } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        location: {
          ...locationInputs[2],
          parentLocationId: 'hncu4u58uxy8y89',
        },
      },
    };
    const { errors, data } = await mutate({
      mutation: CREATE_LOCATION_MUTATION,
      ...payload,
    });
    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors).toMatchSnapshot();
  });
});
