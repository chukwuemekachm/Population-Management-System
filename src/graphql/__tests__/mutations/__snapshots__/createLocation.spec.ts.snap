// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create location mutation should return a location exists error 1`] = `
Array [
  [GraphQLError: A location with the name: Abuja already exists],
]
`;

exports[`create location mutation should return an invalid parentId error 1`] = `
Array [
  [GraphQLError: The parent location with id: hncu4u58uxy8y89 does not exist],
]
`;

exports[`create location mutation should return newly created location 1`] = `
Object {
  "createLocation": Object {
    "femalePopulation": 6474,
    "locationName": "Abuja",
    "malePopulation": 8494,
  },
}
`;

exports[`create location mutation should return validation errors when inputs are not provided 1`] = `
Array [
  [GraphQLError: Some field(s) are failing validation],
]
`;
